### ISSUE TITLE
[JENKINS-26583] Env-inject plugin suppress variables contributed by extension points

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/envinject-plugin#508]
<p>Whenever envvar is available on slave (either inherited from system or configured on slave), its value can not be overridden during build by envvar contributing extension points:</p>

<ul class="alternate" type="square">
	<li><tt>EnvironmentContributor</tt></li>
	<li><tt>BuildWrapper</tt></li>
	<li>possibly others</li>
</ul>

<p>This works correctly when envinject is <em>not</em> installed. The culprit seemsto be that <tt>EnvInjectPluginAction</tt> retains <em>all</em> envvars (not only those envinject contributed) early in the build and put this values back as it implements <tt>EnvironmentContributingAction</tt>. Any value that was overridden by a plugin is then reset to its original value.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivergondza">olivergondza</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583" target="_blank">Env-inject plugin suppress variables contributed by extension points</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: envinject-plugin
<li><b>label(s)</b>: compatibility, environment
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-10-07T19:37:02+00:00
<li><b>votes</b>: 47
<li><b>watchers</b>: 66
<li><b>imported</b>: 20251223-165436
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Whenever envvar is available on slave (either inherited from system or configured on slave), its value can not be overridden during build by envvar contributing extension points:</p>

<ul class="alternate" type="square">
	<li><tt>EnvironmentContributor</tt></li>
	<li><tt>BuildWrapper</tt></li>
	<li>possibly others</li>
</ul>

<p>This works correctly when envinject is <em>not</em> installed. The culprit seemsto be that <tt>EnvInjectPluginAction</tt> retains <em>all</em> envvars (not only those envinject contributed) early in the build and put this values back as it implements <tt>EnvironmentContributingAction</tt>. Any value that was overridden by a plugin is then reset to its original value.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nmanos.png" title="nmanos's avatar" /> nmanos:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=260623&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-260623">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Still relevant on Jenkins 2.7</p>
</pre>
</details>

</ul>
</details>

<p>Still relevant on Jenkins 2.7</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=270865&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270865">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>same here
I confirm that the EnvInjectPluginAction.buildEnvVars() overrides all enviroment variables contribuited to the context (for example by NodeJS plugin in master code in a SimpleBuildWrapper) in the AbstractBuild.getEnvironment() method when cycles on actions.</p>

<p>So the PATH variable contributed by context.env("PATH+NODEJS", path) is override with old one. Now this is blocking because no command are available in the shell resetting any action of the plugin.</p>
</pre>
</details>

</ul>
</details>

<p>same here<br/>
I confirm that the EnvInjectPluginAction.buildEnvVars() overrides all enviroment variables contribuited to the context (for example by NodeJS plugin in master code in a SimpleBuildWrapper) in the AbstractBuild.getEnvironment() method when cycles on actions.</p>

<p>So the PATH variable contributed by context.env("PATH+NODEJS", path) is override with old one. Now this is blocking because no command are available in the shell resetting any action of the plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=284572&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284572">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>any news on this??</p>
</pre>
</details>

</ul>
</details>

<p>any news on this??</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=284638&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284638">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Nikolas Falco
Path:
 src/main/java/jenkins/plugins/nodejs/NodeJSBuildWrapper.java
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java
 src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstallation.java
 src/main/java/jenkins/plugins/nodejs/tools/pathresolvers/FixEnvVarEnvironmentContributingAction.java
<a href="http://jenkins-ci.org/commit/nodejs-plugin/06c1c1b56865bdf93dad79b5440e822ec680191b" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/06c1c1b56865bdf93dad79b5440e822ec680191b</a>
Log:
  Workaround for <a href="https://issues.jenkins.io/browse/JENKINS-26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="JENKINS-26583"><del>JENKINS-26583</del></a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Nikolas Falco<br/>
Path:<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSBuildWrapper.java<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java<br/>
 src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstallation.java<br/>
 src/main/java/jenkins/plugins/nodejs/tools/pathresolvers/FixEnvVarEnvironmentContributingAction.java<br/>
<a href="http://jenkins-ci.org/commit/nodejs-plugin/06c1c1b56865bdf93dad79b5440e822ec680191b" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/06c1c1b56865bdf93dad79b5440e822ec680191b</a><br/>
Log:<br/>
  Workaround for <a href="https://issue-redirect.jenkins.io/issue/26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26583">JENKINS-26583</a>"><del>JENKINS-26583</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=284864&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284864">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Nikolas Falco
Path:
 src/main/java/jenkins/plugins/nodejs/NodeJSBuildWrapper.java
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java
 src/main/java/jenkins/plugins/nodejs/tools/pathresolvers/FixEnvVarEnvironmentContributingAction.java
<a href="http://jenkins-ci.org/commit/nodejs-plugin/dfd4018b3202898d8a20207df996a5ae111bcd1a" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/dfd4018b3202898d8a20207df996a5ae111bcd1a</a>
Log:
  Workaround for <a href="https://issues.jenkins.io/browse/JENKINS-26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="JENKINS-26583"><del>JENKINS-26583</del></a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Nikolas Falco<br/>
Path:<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSBuildWrapper.java<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java<br/>
 src/main/java/jenkins/plugins/nodejs/tools/pathresolvers/FixEnvVarEnvironmentContributingAction.java<br/>
<a href="http://jenkins-ci.org/commit/nodejs-plugin/dfd4018b3202898d8a20207df996a5ae111bcd1a" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/dfd4018b3202898d8a20207df996a5ae111bcd1a</a><br/>
Log:<br/>
  Workaround for <a href="https://issue-redirect.jenkins.io/issue/26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26583">JENKINS-26583</a>"><del>JENKINS-26583</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=285291&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285291">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Nikolas Falco
Path:
 pom.xml
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java
 src/main/java/jenkins/plugins/nodejs/NodeJSPlugin.java
 src/main/java/jenkins/plugins/nodejs/configfiles/NPMConfig.java
 src/main/java/jenkins/plugins/nodejs/configfiles/NPMRegistry.java
 src/main/java/jenkins/plugins/nodejs/configfiles/Npmrc.java
 src/main/java/jenkins/plugins/nodejs/tools/CPU.java
 src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstallation.java
 src/main/resources/lib/nodejs/select.jelly
 src/test/java/jenkins/plugins/nodejs/NodeJSBuildWrapperTest.java
 src/test/java/jenkins/plugins/nodejs/configfiles/NPMConfigTest.java
 src/test/java/jenkins/plugins/nodejs/configfiles/NPMConfigValidationTest.java
<a href="http://jenkins-ci.org/commit/nodejs-plugin/bf452be060c0a723991de990ac6833181ef6089d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/bf452be060c0a723991de990ac6833181ef6089d</a>
Log:
  Merge branch 'master' into <a href="https://issues.jenkins.io/browse/JENKINS-26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="JENKINS-26583"><del>JENKINS-26583</del></a></p>


<p>Compare: <a href="https://github.com/jenkinsci/nodejs-plugin/compare/2d08527b32e7...bf452be060c0" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/nodejs-plugin/compare/2d08527b32e7...bf452be060c0</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Nikolas Falco<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSCommandInterpreter.java<br/>
 src/main/java/jenkins/plugins/nodejs/NodeJSPlugin.java<br/>
 src/main/java/jenkins/plugins/nodejs/configfiles/NPMConfig.java<br/>
 src/main/java/jenkins/plugins/nodejs/configfiles/NPMRegistry.java<br/>
 src/main/java/jenkins/plugins/nodejs/configfiles/Npmrc.java<br/>
 src/main/java/jenkins/plugins/nodejs/tools/CPU.java<br/>
 src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstallation.java<br/>
 src/main/resources/lib/nodejs/select.jelly<br/>
 src/test/java/jenkins/plugins/nodejs/NodeJSBuildWrapperTest.java<br/>
 src/test/java/jenkins/plugins/nodejs/configfiles/NPMConfigTest.java<br/>
 src/test/java/jenkins/plugins/nodejs/configfiles/NPMConfigValidationTest.java<br/>
<a href="http://jenkins-ci.org/commit/nodejs-plugin/bf452be060c0a723991de990ac6833181ef6089d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/nodejs-plugin/bf452be060c0a723991de990ac6833181ef6089d</a><br/>
Log:<br/>
  Merge branch 'master' into <a href="https://issue-redirect.jenkins.io/issue/26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26583">JENKINS-26583</a>"><del>JENKINS-26583</del></a></p>


<p>Compare: <a href="https://github.com/jenkinsci/nodejs-plugin/compare/2d08527b32e7...bf452be060c0" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/nodejs-plugin/compare/2d08527b32e7...bf452be060c0</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=285294&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285294">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sorry, I've rename personal branch for a dedicated workarounf on NodeJS plugin so JIRA should not bother anymore</p>
</pre>
</details>

</ul>
</details>

<p>Sorry, I've rename personal branch for a dedicated workarounf on NodeJS plugin so JIRA should not bother anymore</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>afischer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=287123&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-287123">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Are there any ideas for fixing this problem? Workarounds in all other plugins are not good ideas... Where in the code we can start inspecting the problem?</p>
</pre>
</details>

</ul>
</details>

<p>Are there any ideas for fixing this problem? Workarounds in all other plugins are not good ideas... Where in the code we can start inspecting the problem?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=287130&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-287130">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>EnvInjectPluginAction#buildEnvVars</p>

<p>I do not know the reason of this action, anyway it save an initial environment and put it back After that all buildwrappeer has contribuire the initial environment. So new variables are not affected by this problem, instead changed variables like PATH are restore to its initial value.</p>

<p>Maybe this behaviour / action should be done/registered only if "Override environment variables" is selected in the job configuration page.</p>

<p>There is also a PR in github provides a lot of tests for an eventual solution</p>
</pre>
</details>

</ul>
</details>

<p>EnvInjectPluginAction#buildEnvVars</p>

<p>I do not know the reason of this action, anyway it save an initial environment and put it back After that all buildwrappeer has contribuire the initial environment. So new variables are not affected by this problem, instead changed variables like PATH are restore to its initial value.</p>

<p>Maybe this behaviour / action should be done/registered only if "Override environment variables" is selected in the job configuration page.</p>

<p>There is also a PR in github provides a lot of tests for an eventual solution</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/swf.png" title="swf's avatar" /> swf:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=288358&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288358">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any news on this issue? When will it be fixed? We are currently blocked on updating NodeJS plugin from 0.2.2 to 1.1.0 because it will no longer work together with EnvInject plugin. So right now we are stuck on NodeJS plugin 0.2.2.</p>
</pre>
</details>

</ul>
</details>

<p>Any news on this issue? When will it be fixed? We are currently blocked on updating NodeJS plugin from 0.2.2 to 1.1.0 because it will no longer work together with EnvInject plugin. So right now we are stuck on NodeJS plugin 0.2.2.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> olivergondza:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=288361&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288361">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I did not managed to do a real progress on this thing and I do not see I will have time to get back to this any time soon.</p>
</pre>
</details>

</ul>
</details>

<p>I did not managed to do a real progress on this thing and I do not see I will have time to get back to this any time soon.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>afischer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=288372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can you describe some good starting points for search and modifications in the source?</p>
</pre>
</details>

</ul>
</details>

<p>Can you describe some good starting points for search and modifications in the source?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> olivergondza:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=288390&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288390">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=afischer" class="user-hover" rel="afischer">afischer</a>, see attached PR.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=afischer" class="user-hover" rel="afischer">afischer</a>, see attached PR.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=288504&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288504">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=afischer" class="user-hover" rel="afischer">afischer</a> see my previous post I wrote the class that cause this issue. The git hub <a href="https://github.com/jenkinsci/envinject-plugin/pull/37" class="external-link" target="_blank" rel="nofollow noopener">PR37</a> it's a good start point with a lot of UnitTest for this issue.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=swf" class="user-hover" rel="swf">swf</a> In nodejs github repo there is a branch named workaround-26583 for this issue. You can make a custom build and move into 1.1 snapshot</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=afischer" class="user-hover" rel="afischer">afischer</a> see my previous post I wrote the class that cause this issue. The git hub <a href="https://github.com/jenkinsci/envinject-plugin/pull/37" class="external-link" target="_blank" rel="nofollow noopener">PR37</a> it's a good start point with a lot of UnitTest for this issue.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=swf" class="user-hover" rel="swf">swf</a> In nodejs github repo there is a branch named workaround-26583 for this issue. You can make a custom build and move into 1.1 snapshot</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=293404&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293404">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This bug it's a real pain in the @$$, I have 8/9 bug open on nodejs component due to this.</p>
</pre>
</details>

</ul>
</details>

<p>This bug it's a real pain in the @$$, I have 8/9 bug open on nodejs component due to this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=315363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-315363">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev" class="user-hover" rel="oleg_nenashev">oleg_nenashev</a> not a big surprise. The PR adds other EnvironmentContributor that overrides the enviroment variable map in a different phase than EnvInjectBuilderContributionAction.</p>

<p>In our organisation to get PATH works we patch the EnvInjectBuilderContributionAction simply with </p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">    @Override
    <span class="code-keyword">public</span> void buildEnvVars(@Nonnull AbstractBuild&lt;?, ?&gt; build, @Nonnull EnvVars env) {
        <span class="code-keyword">if</span> (envMap != <span class="code-keyword">null</span>) {
            <span class="code-keyword">for</span> (<span class="code-object">String</span> key : envMap.keySet()) {
                <span class="code-keyword">if</span> (!env.containsKey(key)) {
                    env.put(key, envMap.get(key));
                }
            }
        }
    }
</pre>
</div></div>

<p>the "if (!env.containsKey(key)) {" do the work, just only 1 test fails</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev" class="user-hover" rel="oleg_nenashev">oleg_nenashev</a> not a big surprise. The PR adds other EnvironmentContributor that overrides the enviroment variable map in a different phase than EnvInjectBuilderContributionAction.</p>

<p>In our organisation to get PATH works we patch the EnvInjectBuilderContributionAction simply with </p>


<pre>
    @​Override
    <span class="code-keyword">public</span> void buildEnvVars(@​Nonnull AbstractBuild<?, ?> build, @​Nonnull EnvVars env) {
<span class="code-keyword">if</span> (envMap != <span class="code-keyword">null</span>) {
    <span class="code-keyword">for</span> (<span class="code-object">String</span> key : envMap.keySet()) {
<span class="code-keyword">if</span> (!env.containsKey(key)) {
    env.put(key, envMap.get(key));
}
    }
}
    }
</pre>

<p>the "if (!env.containsKey(key)) {" do the work, just only 1 test fails</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=316235&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316235">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Oliver Gondža
Path:
 src/test/java/org/jenkinsci/plugins/envinject/EnvInjectActionTest.java
<a href="http://jenkins-ci.org/commit/envinject-plugin/1ab1de604bb3c3b7e291d3c581553cb75827bfbb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/envinject-plugin/1ab1de604bb3c3b7e291d3c581553cb75827bfbb</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="JENKINS-26583"><del>JENKINS-26583</del></a> More tests</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Oliver Gondža<br/>
Path:<br/>
 src/test/java/org/jenkinsci/plugins/envinject/EnvInjectActionTest.java<br/>
<a href="http://jenkins-ci.org/commit/envinject-plugin/1ab1de604bb3c3b7e291d3c581553cb75827bfbb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/envinject-plugin/1ab1de604bb3c3b7e291d3c581553cb75827bfbb</a><br/>
Log:<br/>
  <a href="https://issue-redirect.jenkins.io/issue/26583" title="Env-inject plugin suppress variables contributed by extension points" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26583">JENKINS-26583</a>"><del>JENKINS-26583</del></a> More tests</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=316236&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316236">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Oliver Gondža
Path:
 src/main/java/org/jenkinsci/plugins/envinject/EnvInjectListener.java
 src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java
 src/test/java/org/jenkinsci/plugins/envinject/EnvInjectActionTest.java
<a href="http://jenkins-ci.org/commit/envinject-plugin/2948c711b05b68c6a94c334c26fe4e6fdcecda9f" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/envinject-plugin/2948c711b05b68c6a94c334c26fe4e6fdcecda9f</a>
Log:
  <span class="error">&#91;FIXED JENKINS-26583&#93;</span> Do not capture slave variables</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Oliver Gondža<br/>
Path:<br/>
 src/main/java/org/jenkinsci/plugins/envinject/EnvInjectListener.java<br/>
 src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java<br/>
 src/test/java/org/jenkinsci/plugins/envinject/EnvInjectActionTest.java<br/>
<a href="http://jenkins-ci.org/commit/envinject-plugin/2948c711b05b68c6a94c334c26fe4e6fdcecda9f" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/envinject-plugin/2948c711b05b68c6a94c334c26fe4e6fdcecda9f</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED <a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/26583">JENKINS-26583</a>&#93;</span> Do not capture slave variables</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=316237&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316237">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I hope to release the fix on this weekend</p>
</pre>
</details>

</ul>
</details>

<p>I hope to release the fix on this weekend</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>apxeolog:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-26583?focusedId=319766&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-319766">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There is still an issue with plugin overriding env variables even if it is disabled (meaning not enabled in the project configuration). And PATH property still fully changed. I suggest to merge PATH property instead of overriding it (split both PATH's to elements and merge into Set)</p>
</pre>
</details>

</ul>
</details>

<p>There is still an issue with plugin overriding env variables even if it is disabled (meaning not enabled in the project configuration). And PATH property still fully changed. I suggest to merge PATH property instead of overriding it (split both PATH's to elements and merge into Set)</p>
